### PR TITLE
Support casting in Any

### PR DESCRIPTION
**Pull request by lisitsyn**

### CODE REVIEW

**Comment by gf712** [comment]
for the class hierarchy could you generate lots of template specialisations with ctags? and then recursively get the type from the overloaded structs?
```cpp
template <typename T>
struct get_hierarchy {};

template <>
struct get_hierarchy<GaussianKernel>
{
  using types = Types<Kernel, SGObject>;
}
...
```

**Comment by lisitsyn** [comment]
@gf712 nooo not in this PR please 

:)

**Comment by lisitsyn** [comment]
I believe this can't break anything thus merging

**Comment by gf712** [comment]
@lisitsyn Yea, just wondering what you thought about something like that

**Comment by lisitsyn** [comment]
@gf712 looks like a good solution indeed

`std::bases` did not happen :( Sad!

File: src/shogun/lib/any.h [hunk 1]
```diff
@@ -1112,18 +1137,56 @@ namespace shogun {
...
+		static CastingRegistry casting_registry;
 	};
 
 	bool operator==(const Any& lhs, const Any& rhs);
 
 	bool operator!=(const Any& lhs, const Any& rhs);
 
+	template <typename T>
+	inline void register_casts() {
+		if constexpr (std::is_base_of<CSGObject, typename std::remove_pointer<T>::type>::value)
```
Comment (karlnapf) [inline src/shogun/lib/any.h:1178]: Could this be done for all base classes automatically?

**Review by karlnapf — COMMENTED** [review]
Nice! :)

File: src/shogun/lib/any.h (hunk 1 above)
Comment (karlnapf) [reply src/shogun/lib/any.h:1178]: As we have them in is_sg_base already anyways

**Comment by gf712** [comment]
Yea, doesn’t seem like it’ll be around until C++23...

**Comment by vigsterkr** [comment]
for the base classes i think there's one nice crazy fold expression based solution for this.... i'll try later today

**Pull request merged**